src/Internal/DependencyManagement/ExtendedContainer.php

```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
 	 * @throws ContainerException Invalid parameters.
 	 */
 	public function add( string $class_name, $concrete = null, bool $shared = null ) : DefinitionInterface {
-		if ( ! $this->class_is_in_root_namespace( $class_name ) && ! in_array( $class_name, $this->registration_whitelist, true ) ) {
-			throw new ContainerException( "Can't use the container to register '$class_name', only objects in the " . Container::WOOCOMMERCE_ROOT_NAMESPACE . ' namespace are allowed for registration.' );
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "You cannot add '$class_name', only classes in the {$this->woocommerce_namespace} namespace are allowed." );
 		}
 
-		return parent::add( $class_name, $concrete, $shared );
-	}
+		$concrete_class = $this->get_class_from_concrete( $concrete );
+		if ( isset( $concrete_class ) && ! $this->is_class_allowed( $concrete_class ) ) {
+			throw new ContainerException( "You cannot add concrete '$concrete_class', only classes in the {$this->woocommerce_namespace} namespace are allowed." );
+		}
 
-	/**
-	 * Does a class belong to the WooCommerce root namespace?
-	 *
-	 * @param string $class_name The class name to check.
-	 *
-	 * @return bool True if the class belongs to the WooCommerce root namespace.
-	 */
-	private function class_is_in_root_namespace( $class_name ) {
-		return substr( $class_name, 0, strlen( Container::WOOCOMMERCE_ROOT_NAMESPACE ) + 1 ) === Container::WOOCOMMERCE_ROOT_NAMESPACE . '\\';
+		// We want to use a definition class that does not support constructor injection to avoid accidental usage.
+		if ( ! $concrete instanceof DefinitionInterface ) {
+			$concrete = new Definition( $class_name, $concrete );
+		}
```

Comment:
There's a "backdoor" here: an instance of `\League\Container\Definition\Definition` could be passed as a concrete, and this would allow to happily register a class with constructor injection. So better something like this:

```
if ( ! $concrete instanceof DefinitionInterface ) {
    $concrete = new Definition( $class_name, $concrete );
}
else if( ! $concrete instanceof Definition ) {
    $concrete = new Definition( $concrete->getAlias(), $concrete->getConcrete() );
}
```

Reply:
My concern with checking for the base `Definition` is that cases where it is passed might also include arguments, inflectors, and other things that we aren't actually cloning here. Given that they could also implement `DefinitionInterface` in a custom class if they wanted to get around this limitation, my thinking is that we can consider this case the same as we do passing a closure that returns an invalid class.

Reply:
Agreed. We are going to reject classes with constructor arguments anyway so not a big issue.